src/bytecode.c

```diff
@@ -0,0 +1,292 @@
+
+  KBC files are defined in two sections: the code section and the data section.
+  
+    The code section is where all of the linear instructions are set out with
+    jump paths being defined by the compiler using the higher-level language's
+    labels assigned to internal addresses. This is where the result of the
+    compiler goes basically
+
+    The data section is a series of tables pointing to filenames. The compiler
+    defines when a given table should be loaded into memory and the engine loads
+    those assets into the Global Asset Table (GAT). When a given area of the game
+    wants to use those assets, the compiler selects them just based on an offset.
+    mneumonic pseudocode for this might look like
+      LOAD SCENE01   ;; load the assets for scene 1 into the GAT
+      DISP 04        ;; display the 5th entry in the GAT
+      UNLOAD         ;; free the GAT's scene section.
+
+  Conditional options (decisions) are stored as integers (which the user selected)
+  at an offset remembered by the compiler based on labels put in by the programmer.
+  At any time, the comparison instruction can go in and branch based on previous
+  decisions. This is called the Decision Offset Table (DOT).
+
+  For interpreter implementation reasons, all opcodes increment from 0 because they
+  have an O(1) lookup time.
+
+  INSTRUCTION SET
+  ===============
+  00 - Depending on context can either end a string or can be ignored as NOOP.
+  01 - Begin data section
+  02 - end data section
+  03 - begin code section.
+  -- CODE SECTION SPECIFIC --
+  00 - unconditional jump (takes 4 bytes as an operand -- code spot to jump to)
+  01 - End script (return to engine execution)
+  02 - jump if comparison flag is true (takes 4 bytes as an operand)
+  03 - display text at GAT offset (takes 5 bytes as operand 2:offset;2:length;1:flags)
+  04 - display image at GAT offset (takes 5 bytes as operand 2:offset;2:length;1:flags)
+  05 - play sound at GAT offset (takes 5 bytes as operand 2:offset;2:length;1:flags)
+  06 - get user input on option (takes n bytes as input with 2 for each GAT offset)
+       max text length is defined as MAX_OPT_LENGTH in bytecode.h
+  07 - after 2 bytes in a user option indicates a new optionn
+  08 - compare an option at DOT offset to a value (4 bytes -- offset;value)
+  09 - Load assets from x until y into AOT (8 bytes -- x;y)
+  -- DATA SECTION SPECIFIC --
+  00 - begin asset set
+  01 - end asset set
+  02 - load string into current asset set (continues reading until null-byte)
+  03 - load png file based on filename (continues reading until null-byte)
+  04 - load ogg file based on filename (continues reading until null-byte)
+  
+
+  An example bytecode script:
+  01 11 6D 69 61 2e 70 6E 67 00 10 74 65 73 74 20 73 74 72 69 6E 67 00 10
+  6F 70 74 69 6F 6E 31 00 10 6F 70 74 69 6F 6E 32 00 02 03 21 00 01 00 16
+  10 00 01 00 11 00 00 00 1F 00 02 30 00 03 04 00 00 00 00 00 00 00 00 00
```

Comment:
The bytecode reference is out of date, and doesn't fully explain the instruction format